### ISSUE TITLE
formatOnSave not working

### ISSUE THREAD

**Issue by uphlewis** [valeryan/vscode-phpsab#32]
Manually running Format Document works just fine, but with editor.formatOnSave enabled, when saving the formatter doesn't run and i get the following message in the window gutter, but nothing in the extension host log:

![image](https://user-images.githubusercontent.com/43346009/92951934-13382480-f457-11ea-80b4-b9a8f10b78f8.png)

I have debug enabled, but cant see where the debug output is or should be appearing?

**Comment by uphlewis** [comment]
Actually this could be related to #31 - if i change `"editor.formatOnSaveMode": "modifications",`  to `"editor.formatOnSaveMode": "file",` it does appear to work on save.

**Comment by uphlewis** [comment]
In case it helps, i get the same error message appearing in the gutter when highlighting a block of code and doing "Format Selection" regardless of my `"editor.formatOnSaveMode"` setting.

**Comment by valeryan** [comment]
I will check on it this weekend. I have not used this mode in a while because I work in projects that have lots of dirty code and it can be annoying when you need to make a quick hotfix to a file. Probably some changes to the way things are implemented.

**Comment by uphlewis** [comment]
Thanks @valeryan, please let me know if you need any more info (settings/extensions etc)!

**Comment by valeryan** [comment]
I don't implement any format selection functionality currently in this extension. If you have the Format Selection option it is not provided by this extension. I am working on adding support for the format only changed lines option though. If that is what you are looking for we can close this and track it in #31. Thanks.

**Comment by uphlewis** [comment]
Are these not doing the same thing- formatting a subset of the file?

**Comment by edysmp** [comment]
I think they are the same thing. See https://github.com/microsoft/vscode/issues/44075#issuecomment-690990599

**Comment by valeryan** [comment]
Right my extension only currently uses `registerDocumentFormattingEditProvider` to add the phpcbf fixing for a whole file. To support the format selection and the new format only modified lines on save I will need to implement 'registerDocumentRangeFormattingEditProvider' and rework the fixer class to handle only pieces of a document.

**Comment by valeryan** [comment]
TLDR: phpcbf needs to be aware of the whole file to apply rules correctly and format by selection is a bad idea. 

The problem with providing format selection and format only modified is managing whitespace. phpcbf rules that manage whitespace can freak out when you only try to format partial code. I could use the phpcbf option `--filter=gitmodified` to accomplish only formatting modified lines on saving but because the two operations (format selection, format only modified) use the same hooks I cannot support this without enabling format selection. That would mean I have to write a bunch of logic to either expand the selection range based on the rule that is being applied or write rules to fix whitespace discrepancies that occur. 

So currently supporting this as vscode as implemented it is not feasible for me.

Another problem with the implementation is that because I do not implement `registerDocumentRangeFormattingEditProvider` my code is not called when the vscode option of format only modified is not called in any way so I can not even provide an error message saying this is not implemented without also enabling the menu option for format selection which I think will cause confusion for it to show up only to error and tell you that this functionality is not implemented.

**Comment by valeryan** [comment]
I have implemented `registerDocumentRangeFormattingEditProvider` and use it to trigger phpcbf to run with the git modified filter when format on save mode is set to format modified.
